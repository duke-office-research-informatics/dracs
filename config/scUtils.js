import React from "react";
import { ThemeProvider } from "styled-components";
import { mount, shallow } from "enzyme";
import { theme as defaultTheme } from "../lib/dracs.es.js";

export const shallowWithTheme = (children, customTheme) => {
  const theme = customTheme || defaultTheme;
  return shallow(children, { context: theme });
};

/* eslint-disable */
export function mountWithTheme(child) {
  return mount(child, {
    wrappingComponent: ({ children }) => (
      <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
    ),
  });
}

//FOR ACCESSING PORTALS VIA ENZYME -- import portal from dracs
// https://github.com/airbnb/enzyme/issues/1202#issuecomment-351639783
// function mountPortal(el){
//   const wrapper =  mountWithTheme(el);
//   wrapper.setState({ visible: true});
//   const portalInstance = wrapper.find(Portal).instance();
//   return mountWithTheme(portalInstance.props.children);
// }

//FOR ACCESSING EVENT LISTENERS ATTACHED TO DOCUMENT OR DOCUMENT.BODY
//https://github.com/airbnb/enzyme/issues/426#issuecomment-296304015
// const map = {};
// document.body.addEventListener = jest.fn((event, cb) => {
//   map[event] = cb;
// });
