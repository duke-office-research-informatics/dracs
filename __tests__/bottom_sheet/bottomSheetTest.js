import React from "react";
import { mountWithTheme } from "../../config/scUtils.js";
import { axe } from "../../config/axe-helper.js";
import { theme, BottomSheet, IconMenu } from "../../lib/dracs.es.js";

const clickFn = jest.fn();

describe("BottomSheet", () => {
  describe("`active` false", () => {
    it("matches snapshot", () => {
      const wrapper = mountWithTheme(<BottomSheet />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("`active` true", () => {
    it("matches snapshot", () => {
      const wrapper = mountWithTheme(<BottomSheet active={true} />);
      expect(wrapper).toMatchSnapshot();
    });
    it("returns no a11y violations", async () => {
      const wrapper = mountWithTheme(<BottomSheet active={true} />);
      const results = await axe(wrapper.html());
      expect(results).toHaveNoViolations();
    });
    it("renders an overlay and a body", () => {
      const wrapper = mountWithTheme(<BottomSheet active={true} />);
      expect(wrapper.find("bottom_sheet__SheetOverlay").length).toBe(1);
      expect(wrapper.find("bottom_sheet__SheetBody").length).toBe(1);
    });
    it("applies the correct styles to the sheet overlay", () => {
      const wrapper = mountWithTheme(<BottomSheet active={true} />);
      wrapper.setState({ visible: true });
      const div = wrapper.find("bottom_sheet__SheetOverlay");
      expect(div).toHaveStyleRule("position", "absolute");
      expect(div).toHaveStyleRule("top", "0");
      expect(div).toHaveStyleRule("left", "0");
      expect(div).toHaveStyleRule("bottom", "0");
      expect(div).toHaveStyleRule("right", "0");
      expect(div).toHaveStyleRule("background-color", "rgba(0,0,0,0.1)");
      expect(div).toHaveStyleRule("z-index", `${theme.zIndex.xl}`);
      expect(div).toHaveStyleRule("overflow-y", "hidden");
    });
    it("applies the correct styles to the sheet body", () => {
      const wrapper = mountWithTheme(<BottomSheet active={true} />);
      wrapper.setState({ visible: true });
      const div = wrapper.find("bottom_sheet__SheetBody");
      expect(div).toHaveStyleRule("position", "absolute");
      expect(div).toHaveStyleRule("bottom", "0");
      expect(div).toHaveStyleRule("box-sizing", "border-box");
      expect(div).toHaveStyleRule("background-color", "#fff");
      expect(div).toHaveStyleRule(
        "border-radius",
        `${theme.sizing.borderRadius}px ${theme.sizing.borderRadius}px 0 0`
      );
      expect(div).toHaveStyleRule("min-height", "36px");
      expect(div).toHaveStyleRule("transform", "translate3d(0,0,0)");
      expect(div).toHaveStyleRule(
        "transition",
        "all 300ms cubic-bezier(0.4,0,0.2,1)"
      );
      expect(div).toHaveStyleRule("overflow", "hidden");
    });

    describe("`bgColor` prop", () => {
      it("matches snapshot", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} bgColor={theme.colors.action} />
        );
        wrapper.setState({ visible: true });
        expect(wrapper).toMatchSnapshot();
      });
      it("applies a bg color to the sheet body", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} bgColor={theme.colors.action} />
        );
        wrapper.setState({ visible: true });
        const div = wrapper.find("bottom_sheet__SheetBody");
        expect(div).toHaveStyleRule("background-color", theme.colors.action);
      });
    });

    describe("`bodyStyle`", () => {
      it("matches snapshot", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} bodyStyle={{ maxHeight: "100px" }} />
        );
        wrapper.setState({ visible: true });
        expect(wrapper).toMatchSnapshot();
      });
      it("applies the style object to the sheet body", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} bodyStyle={{ maxHeight: "100px" }} />
        );
        wrapper.setState({ visible: true });
        const div = wrapper.find("bottom_sheet__SheetBody");
        expect(div.props().style).toHaveProperty("maxHeight", "100px");
      });
    });

    describe("`children`", () => {
      it("matches snapshot", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true}>
            <IconMenu />
          </BottomSheet>
        );
        wrapper.setState({ visible: true });
        expect(wrapper).toMatchSnapshot();
      });
      it("renders the child in the sheet body", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true}>
            <IconMenu />
          </BottomSheet>
        );
        wrapper.setState({ visible: true });
        const child = wrapper.find("bottom_sheet__SheetBody").find(IconMenu);
        expect(child.length).toBe(1);
      });
    });

    describe("`delay`", () => {
      it("matches snapshot", () => {
        const wrapper = mountWithTheme(<BottomSheet active={true} delay={0} />);
        wrapper.setState({ visible: true });
        expect(wrapper).toMatchSnapshot();
      });
      it("applies the delay to the body transition", () => {
        const wrapper = mountWithTheme(<BottomSheet active={true} delay={0} />);
        wrapper.setState({ visible: true });
        const div = wrapper.find("bottom_sheet__SheetBody");
        expect(div).toHaveStyleRule(
          "transition",
          "all 0ms cubic-bezier(0.4,0,0.2,1)"
        );
      });
    });

    describe("`height`", () => {
      it("matches snapshot", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} height="10em" />
        );
        wrapper.setState({ visible: true });
        expect(wrapper).toMatchSnapshot();
      });
      it("applies height to the sheet body", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} height="10em" />
        );
        wrapper.setState({ visible: true });
        const div = wrapper.find("bottom_sheet__SheetBody");
        expect(div).toHaveStyleRule("height", "10em");
      });
    });

    describe("`onOverlayClick`", () => {
      it("matches snapshot", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} onOverlayClick={clickFn} />
        );
        wrapper.setState({ visible: true });
        expect(wrapper).toMatchSnapshot();
      });
      it("executes click handler", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} onOverlayClick={clickFn} />
        );
        wrapper.setState({ visible: true });
        wrapper.find("bottom_sheet__SheetOverlay").simulate("click");
        expect(clickFn).toBeCalled();
      });
      // it('does not trigger click handler on cick of body', () => {
      //   const wrapper = mountWithTheme(<BottomSheet active={true} onOverlayClick={clickFn}/>);
      //   wrapper.setState({visible: true});
      //   wrapper.find('bottom_sheet__SheetBody').simulate('click');
      //   expect(clickFn).not.toBeCalled();
      // })
    });

    describe("`overlayStyle`", () => {
      it("matches snapshot", () => {
        const wrapper = mountWithTheme(
          <BottomSheet
            active={true}
            overlayStyle={{ border: `solid 1px ${theme.colors.border}` }}
          />
        );
        wrapper.setState({ visible: true });
        expect(wrapper).toMatchSnapshot();
      });
      it("applies the style object to the sheet overlay", () => {
        const wrapper = mountWithTheme(
          <BottomSheet
            active={true}
            overlayStyle={{ border: `solid 1px ${theme.colors.border}` }}
          />
        );
        wrapper.setState({ visible: true });
        const div = wrapper.find("bottom_sheet__SheetOverlay");
        expect(div.props().style).toHaveProperty(
          "border",
          `solid 1px ${theme.colors.border}`
        );
      });
    });

    describe("`textColor`", () => {
      it("matches snapshot", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} textColor="#fff" />
        );
        wrapper.setState({ visible: true });
        expect(wrapper).toMatchSnapshot();
      });
      it("applies the text color to the sheet body", () => {
        const wrapper = mountWithTheme(
          <BottomSheet active={true} textColor="#fff" />
        );
        wrapper.setState({ visible: true });
        const div = wrapper.find("bottom_sheet__SheetBody");
        expect(div).toHaveStyleRule("color", "#fff");
      });
    });
  });
});
