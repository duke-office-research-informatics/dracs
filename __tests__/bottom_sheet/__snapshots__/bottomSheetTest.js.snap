// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheet \`active\` false matches snapshot 1`] = `""`;

exports[`BottomSheet \`active\` true \`bgColor\` prop matches snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  background-color: #0680cd;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  />
</div>
`;

exports[`BottomSheet \`active\` true \`bodyStyle\` matches snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
    style={
      Object {
        "maxHeight": "100px",
      }
    }
  />
</div>
`;

exports[`BottomSheet \`active\` true \`children\` matches snapshot 1`] = `
.c2 {
  fill: #fff;
  vertical-align: top;
}

.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  >
    <svg
      aria-label="open customize and profile menu"
      className="c2"
      color="#fff"
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
    >
      <title>
        menu icon
      </title>
      <desc>
        hamburger menu icon
      </desc>
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
      />
    </svg>
  </div>
</div>
`;

exports[`BottomSheet \`active\` true \`delay\` matches snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: all 0ms cubic-bezier(0.4,0,0.2,1);
  transition: all 0ms cubic-bezier(0.4,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  />
</div>
`;

exports[`BottomSheet \`active\` true \`height\` matches snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  height: 10em;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1"
    height="10em"
    onClick={[Function]}
  />
</div>
`;

exports[`BottomSheet \`active\` true \`onOverlayClick\` matches snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
  onClick={[Function]}
>
  <div
    className="c1"
    onClick={[Function]}
  />
</div>
`;

exports[`BottomSheet \`active\` true \`overlayStyle\` matches snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
  style={
    Object {
      "border": "solid 1px #e8e5e2",
    }
  }
>
  <div
    className="c1"
    onClick={[Function]}
  />
</div>
`;

exports[`BottomSheet \`active\` true \`textColor\` matches snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  color: #fff;
  background-color: #fff;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  transition: all 300ms cubic-bezier(0.4,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  />
</div>
`;

exports[`BottomSheet \`active\` true matches snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0,0,0,0.1);
  z-index: 1000;
  overflow-y: hidden;
}

.c1 {
  position: absolute;
  bottom: 0;
  width: 100%;
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 4px 4px 0 0;
  min-height: 36px;
  -webkit-transform: translate3d(0,100%,0);
  -ms-transform: translate3d(0,100%,0);
  transform: translate3d(0,100%,0);
  -webkit-transition: all 300ms cubic-bezier(0,0,0.2,1);
  transition: all 300ms cubic-bezier(0,0,0.2,1);
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  />
</div>
`;
